# User Guide

DCLS GDIP runs from the Linux command line and consists of a number of shell scripts and a single Makefile compiling them all. Default parameters will generate a report for NGS reads obtained from *Salmonella enterica* subsp enteritidis isolates (Reference strain: [P125109 (NC_011294.1)](http://www.ncbi.nlm.nih.gov/nuccore/NC_011294.1), MLST; [Salmonella enterica](http://mlst.warwick.ac.uk/mlst/dbs/Senterica/Downloads_HTML)). For other species, see [Advanced usage](#advanced-usage).  

This guide assumes an AWS EC2 instance launched from the DCLS GDIP image ami-9f630088.

# Basic usage

## Input

DCLS GDIP assumes a project directory with paired fastq files ending in either: 

- `_1.fastq.gz` and `_2.fastq.gz`; or 
- `R1.fastq.gz` and `R2.fastq.gz`

If your fastq.gz files are on a local drive, create a new project directory on your EC2 instance and use `scp` to copy sequence files into that directory:

- From the EC2 instance: `mkdir MyProjectDir`
- From your local drive: `scp ~/MyLocalSeqDir/*fastq.gz ubuntu@InstanceIP:~/MyProjectDir/`

Alternatively, use a client such as [Cyberduck](https://cyberduck.io) to open up a transfer window, or [Mountainduck](https://mountainduck.io/) to mount a cloud storage drive as a local disk in Mac/Finder or Windows/Explorer.


## Setup

The rest of this documentation assumes the [dcls/gdip/scripts](https://github.com/dcls/gdip/tree/master/scripts) directory is on your `$PATH`. If you are using the DCLS GDIP AMI, it already is. If you are not:

```sh
git clone https://github.com/dcls/gdip ~/gdip
export PATH=~/gdip/scripts:$PATH
```

Once a project directory has been created containing the paired-end `_1.fastq.gz`+`_2.fastq.gz` reads of interest, copy the Makefile into your project directory:

```sh
copy_makefile
```

## Types of analysis
 
1. For a full report (SNP matrix, phylogenetic tree, AMR, MLST) run: `make`.
2. For a SNP matrix and a phylogenetic tree run: `make tree-only`.
3. For AMR results only, run: `make amr`.
4. For MLST results only, run: `make mlst`.

## Output

The default DCLS GDIP pipeline (running `make`) will generate a `<DATE>-gdip-report.zip` file containing all requested results. Additional files related to the analysis (e.g. summary logs and intermedaite analysis reports) generated by CFSAN-SNP Pipeline and SRST2 can be found within the CFSAN-output and SRST2 directories, respectively. For more information on other output files, see [CFSAN-SNP Pipeline](http://snp-pipeline.readthedocs.io/en/latest/usage.html) and [SRST2](https://github.com/katholt/srst2) documentaion.

Large intermediate files can be removed to save disk space by running: `make space` after a run has been completed (this removes sam/bam alignment files in the cfsan output directory).

## Transfering data to local drive

The `scp` command can be used to copy the gdip report to your local drive:

```sh
scp ubuntu@InstanceIP:~/MyProjectDir/<DATE>-gdip-report.zip ~/SomeLocalDir/
```

Alternatively, use a client such as [Cyberduck](https://cyberduck.io) to open up a transfer window, or [Mountainduck](https://mountainduck.io/) to mount a cloud storage drive as a local disk in Mac/Finder or Windows/Explorer.

# Advanced Usage

## DCLS GDIP for different species 

Default settings will generate results using genome/MLST references approriate for *Salmonella enterica* subsp Enteritidis. This can be changed by downloading alternative references and redirecting `cfsan-snp.sh` and `mlstsrst2.sh` to those files. Once `cfsan-snp.sh` and `mlstsrst2.sh` have been editted appropriately, DCLS GDIP can be invoked as described above. 

## Change CFSAN-SNP reference

Download alternative reference in .fasta format and edit the Makefile in `~/gdip/scripts`.

Change:

```make
REF_GENOME := /opt/genomes/senteritidisp125109.fasta
```

to:

```make
REF_GENOME := /path/to/alternativeReference.fasta
```

## Change MLST reference

Download fasta allele sequences and a tab-delimited file which defines the ST profiles as a combination of alleles. These can be retrieved automatically from pubmlst.org/data/ using the script provided by SRST2, e.g.

```sh
getmlst.py --species "Escherichia coli"
```

Once these files are retrieved, edit `gdip/scripts/mlstsrst2.sh`:

Change

```sh
MLST_DB="/opt/genomes/mlst_senterica/Salmonella_enterica.fasta"
MLST_DEF="/opt/genomes/mlst_senterica/senterica.txt"
```

to:

```
MLST_DB="/path/to/alternative/mlst/species.fasta"
MLST_DEF="/path/to/alternative/mlst/speciesdef.txt"
```

## Scripts

The Makefile runs several scripts, each performing a piece of the analysis. Scripts can be run individually. View script sources [here](https://github.com/dcls/gdip/blob/master/scripts/Makefile).

|  Script | Discription   | 
| --- | --- |
| [cfsan-prepare-directories.sh](https://github.com/dcls/gdip/blob/master/scripts/cfsan-prepare-directories.sh) | Organizes reads and symbolic links into the correct directory structure as needed by CFSAN SNP pipeline.  |
| [multisrst2.sh](https://github.com/dcls/gdip/blob/master/scripts/multisrst2.sh) |  Runs SRST2 to identify antimicrobial resistance genes on multiple samples in a single directory. Run with `-h` option to get some help. |
| [mlstsrst2.sh](https://github.com/dcls/gdip/blob/master/scripts/mlstsrst2.sh) | Runs SRST2 to generate MLST report on multiple samples in a single directory. Run with `-h` option to get some help. |
| [Makefile](https://github.com/dcls/gdip/blob/master/scripts/Makefile) | Runs the shell scripts listed above, the CFSAN-SNP pipeline and FastTree to generate a final report. |
| [copy_makefile](https://github.com/dcls/gdip/blob/master/scripts/copy_makefile) | Copies the Makefile above to the current working directory, and prints out some instructions. |
