# User Guide

DCLS GDIP runs from the Linux command line and consists of a number of shell scripts and a single Makefile compiling them all. Default parameters will generate a report for NGS reads obtained from *Salmonella enterica* subsp enteritidis isolates (Reference strain: [P125109 (NC_011294.1)](http://www.ncbi.nlm.nih.gov/nuccore/NC_011294.1), MLST; [Salmonella enterica](http://mlst.warwick.ac.uk/mlst/dbs/Senterica/Downloads_HTML)). For other species, see [Advanced usage](#advanced-usage).  

This guide assumes an AWS EC2 instance launched from the DCLS GDIP image ami-XXXXX.

# Quick Start

## Input

DCLS GDIP assumes a project directory with paired fastq files ending in either: 

- `_1.fastq.gz` and `_2.fastq.gz`; or 
- `R1.fastq.gz` and `R2.fastq.gz`

If your fastq.gz files are on a local drive, create a new project directory on your EC2 instance and use `scp` to copy sequence files into that directory:

- From the EC2 instance: `mkdir MyProjectDir`
- From your local drive: `scp ~/MyLocalSeqDir/*fastq.gz ubuntu@InstanceIP:~/MyProjectDir/`

Alternatively, use a client such as [Cyberduck](https://cyberduck.io) to open up a transfer window, or [Mountainduck](https://mountainduck.io/) to mount a cloud storage drive as a local disk in Mac/Finder or Windows/Explorer.


## Set up

Once a project directory has been created containing the pair-end fastq.gz reads of interest, copy the Makefile into your project directory:

`cp ~gdip/scripts/Makefile ~/MyProjectDir`

## Types of analysis
 
**1)** For a full report (i.e a snp-matrix, phylogenetic tree, and both an MLST and AMR results) run: `make` 

**2)** For a snp-matrix and a phylogenetic tree run: `make tree-only`

**3)** For MLST results run: `make mlst`

**4)** For a AMR results run: `make amr`
 
## Output

DCLS GDIP will generate a `<DATE>-gdip-report.zip` file containing all requested results. Additional files related to the analysis (e.g. summary logs and intermedaite analysis reports) generated by CFSAN-SNP Pipeline and SRST2 can be found within the CFSAN-output and SRST2  directories, resepectiely. For more information on other output files, see [CFSAN-SNP Pipeline](http://snp-pipeline.readthedocs.io/en/latest/usage.html) and [SRST2](https://github.com/katholt/srst2) documentaion.

Large intermediate files can be removed to save disk space by running: `make make_space` after a run has been completed

## Transfering data to local drive

The `scp` command can be utilized to copy the gdip report to your local drive:
`scp ubuntu@InstanceIP:~/MyProjectDir/<DATE>-gdip-report.zip ~/SomeLocalDir/`

# Advanced Usage

## DCLS GDIP for different species 

Default settings will generate results using genome/MLST references approriate for *Salmonella enterica* subsp enteritidis reads. This can be changed by downloading alternative references and redirecting `cfsan-snp.sh` and `mlstsrst2.sh` to those files. Once `cfsan-snp.sh` and `mlstsrst2.sh` have been editted appropriately, DCLS GDIP can be invoked as described above. 

## Change CFSAN-SNP reference

Download alternative reference in .fasta format and edit 

`cfsan-snp.sh`:

`vi ~/gdip/scripts/cfsan-snp.sh`

Change 

`REF_GENOME="/opt/genomes/senteritidisp125109.fasta"`

to 

`REF_GENOME="PathToAlternativeReference.fasta"`

and save these changes: 

`esc :wq`

##Change MLST reference

Download fasta allele sequences and a tab-delim file which defines the ST profiles as a combination of alleles. These can be retrieved automatically from pubmlst.org/data/ using the script provided by SRST2, e.g.

`getmlst.py --species "Escherichia coli"`

Once these files are retrieved, edit `mlstsrst2.sh`:

`vi ~/gdip/scripts/mlstsrst2.sh`

Change 

`MLST_DB="/opt/genomes/mlst_senterica/Salmonella_enterica.fasta"`

`MLST_DEF="/opt/genomes/mlst_senterica/senterica.txt"`

to:

`MLST_DB="PathToAlternativeAllelSequences.fasta"`

`MLST_DEF="PathToAlternativeSTDefinitions.txt"`

and save these changes:

`esc :wq`
 

## Scripts
|  Script | Discription   | 
|---|---|
|  cfsan-prepare-directories.sh | Organizes reads into correct directory structure as needed by CFSAN SNP pipeline  |
|  multisrst2.sh |  Runs SRST2 to identify antimicrobial resistance genes |
|  mlstsrst2.sh | Runs SRST2 to generate MLST report  |
|  Makefile | Runs the shell scripts listed above, the CFSAN-SNP pipeline and FastTree to generate a final report  |