<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <title>User Guide - DCLS GDIP</title>
  

  <link rel="shortcut icon" href="../img/favicon.ico">

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../css/highlight.css">

  
  <script>
    // Current page data
    var mkdocs_page_name = "User Guide";
    var mkdocs_page_input_path = "userguide.md";
    var mkdocs_page_url = "/userguide/";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js"></script>
  <script src="../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../js/highlight.pack.js"></script>
  <script src="../js/theme.js"></script> 

  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> DCLS GDIP</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        <ul class="current">
          
            <li>
    <li class="toctree-l1 ">
        <a class="" href="..">Home</a>
        
    </li>
<li>
          
            <li>
    <li class="toctree-l1 current">
        <a class="current" href="./">User Guide</a>
        
            <ul>
            
                <li class="toctree-l3"><a href="#user-guide">User Guide</a></li>
                
            
                <li class="toctree-l3"><a href="#basic-usage">Basic usage</a></li>
                
                    <li><a class="toctree-l4" href="#input">Input</a></li>
                
                    <li><a class="toctree-l4" href="#setup">Setup</a></li>
                
                    <li><a class="toctree-l4" href="#types-of-analysis">Types of analysis</a></li>
                
                    <li><a class="toctree-l4" href="#output">Output</a></li>
                
                    <li><a class="toctree-l4" href="#transfering-data-to-local-drive">Transfering data to local drive</a></li>
                
            
                <li class="toctree-l3"><a href="#advanced-usage">Advanced Usage</a></li>
                
                    <li><a class="toctree-l4" href="#dcls-gdip-for-different-species">DCLS GDIP for different species</a></li>
                
                    <li><a class="toctree-l4" href="#change-cfsan-snp-reference">Change CFSAN-SNP reference</a></li>
                
                    <li><a class="toctree-l4" href="#change-mlst-reference">Change MLST reference</a></li>
                
                    <li><a class="toctree-l4" href="#scripts">Scripts</a></li>
                
            
            </ul>
        
    </li>
<li>
          
            <li>
    <li class="toctree-l1 ">
        <a class="" href="../about/">About</a>
        
    </li>
<li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">DCLS GDIP</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>User Guide</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="user-guide">User Guide</h1>
<p>DCLS GDIP runs from the Linux command line and consists of a number of shell scripts and a single Makefile compiling them all. Default parameters will generate a report for NGS reads obtained from <em>Salmonella enterica</em> subsp enteritidis isolates (Reference strain: <a href="http://www.ncbi.nlm.nih.gov/nuccore/NC_011294.1">P125109 (NC_011294.1)</a>, MLST; <a href="http://mlst.warwick.ac.uk/mlst/dbs/Senterica/Downloads_HTML">Salmonella enterica</a>). For other species, see <a href="#advanced-usage">Advanced usage</a>.  </p>
<p>This guide assumes an AWS EC2 instance launched from the DCLS GDIP image ami-XXXXX.</p>
<h1 id="basic-usage">Basic usage</h1>
<h2 id="input">Input</h2>
<p>DCLS GDIP assumes a project directory with paired fastq files ending in either: </p>
<ul>
<li><code>_1.fastq.gz</code> and <code>_2.fastq.gz</code>; or </li>
<li><code>R1.fastq.gz</code> and <code>R2.fastq.gz</code></li>
</ul>
<p>If your fastq.gz files are on a local drive, create a new project directory on your EC2 instance and use <code>scp</code> to copy sequence files into that directory:</p>
<ul>
<li>From the EC2 instance: <code>mkdir MyProjectDir</code></li>
<li>From your local drive: <code>scp ~/MyLocalSeqDir/*fastq.gz ubuntu@InstanceIP:~/MyProjectDir/</code></li>
</ul>
<p>Alternatively, use a client such as <a href="https://cyberduck.io">Cyberduck</a> to open up a transfer window, or <a href="https://mountainduck.io/">Mountainduck</a> to mount a cloud storage drive as a local disk in Mac/Finder or Windows/Explorer.</p>
<h2 id="setup">Setup</h2>
<p>The rest of this documentation assumes the <a href="https://github.com/dcls/gdip/tree/master/scripts">dcls/gdip/scripts</a> directory is on your <code>$PATH</code>. If you are using the DCLS GDIP AMI, it already is. If you are not:</p>
<pre><code class="sh">git clone https://github.com/dcls/gdip ~/gdip
export PATH=~/gdip/scripts:$PATH
</code></pre>

<p>Once a project directory has been created containing the paired-end <code>_1.fastq.gz</code>+<code>_2.fastq.gz</code> reads of interest, copy the Makefile into your project directory:</p>
<pre><code class="sh">copy_makefile
</code></pre>

<h2 id="types-of-analysis">Types of analysis</h2>
<ol>
<li>For a full report (SNP matrix, phylogenetic tree, AMR, MLST) run: <code>make</code>.</li>
<li>For a SNP matrix and a phylogenetic tree run: <code>make tree-only</code>.</li>
<li>For AMR results only, run: <code>make amr</code>.</li>
<li>For MLST results only, run: <code>make mlst</code>.</li>
</ol>
<h2 id="output">Output</h2>
<p>The default DCLS GDIP pipeline (running <code>make</code>) will generate a <code>&lt;DATE&gt;-gdip-report.zip</code> file containing all requested results. Additional files related to the analysis (e.g. summary logs and intermedaite analysis reports) generated by CFSAN-SNP Pipeline and SRST2 can be found within the CFSAN-output and SRST2 directories, respectively. For more information on other output files, see <a href="http://snp-pipeline.readthedocs.io/en/latest/usage.html">CFSAN-SNP Pipeline</a> and <a href="https://github.com/katholt/srst2">SRST2</a> documentaion.</p>
<p>Large intermediate files can be removed to save disk space by running: <code>make space</code> after a run has been completed (this removes sam/bam alignment files in the cfsan output directory).</p>
<h2 id="transfering-data-to-local-drive">Transfering data to local drive</h2>
<p>The <code>scp</code> command can be used to copy the gdip report to your local drive:</p>
<pre><code class="sh">scp ubuntu@InstanceIP:~/MyProjectDir/&lt;DATE&gt;-gdip-report.zip ~/SomeLocalDir/
</code></pre>

<p>Alternatively, use a client such as <a href="https://cyberduck.io">Cyberduck</a> to open up a transfer window, or <a href="https://mountainduck.io/">Mountainduck</a> to mount a cloud storage drive as a local disk in Mac/Finder or Windows/Explorer.</p>
<h1 id="advanced-usage">Advanced Usage</h1>
<h2 id="dcls-gdip-for-different-species">DCLS GDIP for different species</h2>
<p>Default settings will generate results using genome/MLST references approriate for <em>Salmonella enterica</em> subsp Enteritidis. This can be changed by downloading alternative references and redirecting <code>cfsan-snp.sh</code> and <code>mlstsrst2.sh</code> to those files. Once <code>cfsan-snp.sh</code> and <code>mlstsrst2.sh</code> have been editted appropriately, DCLS GDIP can be invoked as described above. </p>
<h2 id="change-cfsan-snp-reference">Change CFSAN-SNP reference</h2>
<p>Download alternative reference in .fasta format and edit the Makefile in <code>~/gdip/scripts</code>.</p>
<p>Change:</p>
<pre><code class="make">REF_GENOME := /opt/genomes/senteritidisp125109.fasta
</code></pre>

<p>to:</p>
<pre><code class="make">REF_GENOME := /path/to/alternativeReference.fasta
</code></pre>

<h2 id="change-mlst-reference">Change MLST reference</h2>
<p>Download fasta allele sequences and a tab-delimited file which defines the ST profiles as a combination of alleles. These can be retrieved automatically from pubmlst.org/data/ using the script provided by SRST2, e.g.</p>
<pre><code class="sh">getmlst.py --species &quot;Escherichia coli&quot;
</code></pre>

<p>Once these files are retrieved, edit <code>gdip/scripts/mlstsrst2.sh</code>:</p>
<p>Change</p>
<pre><code class="sh">MLST_DB=&quot;/opt/genomes/mlst_senterica/Salmonella_enterica.fasta&quot;
MLST_DEF=&quot;/opt/genomes/mlst_senterica/senterica.txt&quot;
</code></pre>

<p>to:</p>
<pre><code>MLST_DB=&quot;/path/to/alternative/mlst/species.fasta&quot;
MLST_DEF=&quot;/path/to/alternative/mlst/speciesdef.txt&quot;
</code></pre>

<h2 id="scripts">Scripts</h2>
<p>The Makefile runs several scripts, each performing a piece of the analysis. Scripts can be run individually. View script sources <a href="https://github.com/dcls/gdip/blob/master/scripts/Makefile">here</a>.</p>
<table>
<thead>
<tr>
<th>Script</th>
<th>Discription</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://github.com/dcls/gdip/blob/master/scripts/cfsan-prepare-directories.sh">cfsan-prepare-directories.sh</a></td>
<td>Organizes reads and symbolic links into the correct directory structure as needed by CFSAN SNP pipeline.</td>
</tr>
<tr>
<td><a href="https://github.com/dcls/gdip/blob/master/scripts/multisrst2.sh">multisrst2.sh</a></td>
<td>Runs SRST2 to identify antimicrobial resistance genes on multiple samples in a single directory. Run with <code>-h</code> option to get some help.</td>
</tr>
<tr>
<td><a href="https://github.com/dcls/gdip/blob/master/scripts/mlstsrst2.sh">mlstsrst2.sh</a></td>
<td>Runs SRST2 to generate MLST report on multiple samples in a single directory. Run with <code>-h</code> option to get some help.</td>
</tr>
<tr>
<td><a href="https://github.com/dcls/gdip/blob/master/scripts/Makefile">Makefile</a></td>
<td>Runs the shell scripts listed above, the CFSAN-SNP pipeline and FastTree to generate a final report.</td>
</tr>
<tr>
<td><a href="https://github.com/dcls/gdip/blob/master/scripts/copy_makefile">copy_makefile</a></td>
<td>Copies the Makefile above to the current working directory, and prints out some instructions.</td>
</tr>
</tbody>
</table>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../about/" class="btn btn-neutral float-right" title="About">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href=".." class="btn btn-neutral" title="Home"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
	  
        </div>
      </div>

    </section>

  </div>

<div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href=".." style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../about/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>

</body>
</html>
